### PR TITLE
Optimize signextend

### LINKED ISSUE #260: Optimise SIGNEXTEND to use native bn.js and not array functions

### PR DESCRIPTION
**Pull request by axic**
Fixes #260.

Had this in a local branch for a while, need to verify first.

### CODE REVIEW

**Comment by axic** [comment]
@holgerd77 do you want to have a look at this?

**Comment by holgerd77** [comment]
Yes, can do.

**Comment by holgerd77** [comment]
Not working yet:

```
node ./tests/tester -s --test='randomStatetest470'
```

**Comment by axic** [comment]
Here is a bignumber based implementation: https://github.com/ethereum/solidity/blob/iulia_interp/libjulia/interpreter/EVMInstructionInterpreter.cpp#L145-L158

Also here's one for `BYTE`: https://github.com/ethereum/solidity/blob/iulia_interp/libjulia/interpreter/EVMInstructionInterpreter.cpp#L139-L140 (where `arg[1] >>` is a right shift on a big number)

File: lib/opFns.js [hunk 1]
```diff
@@ -102,23 +102,21 @@ module.exports = {
     }
   },
   SIGNEXTEND: function (k, val, runState) {
-    val = val.toArrayLike(Buffer, 'be', 32)
-    var extendOnes = false
-
     if (k.lten(31)) {
```
Comment (axic) [inline lib/opFns.js]: This should be ``ltn(31)` or `lten(30)`.

**Comment by axic** [comment]
(did a merge commit to retrigger tests)

**Comment by holgerd77** [comment]
Rebase/merge on this failed, will omit for now.

**Comment by holgerd77** [comment]
Update: started with a fresh branch worked, so rebased this as well.

**Comment by axic** [comment]
After merging #307 we finally know why this is failing: 

```
/home/circleci/project/ethreumjs-vm/dist/runCode.js:205
          throw e;
          ^

TypeError: mask.inot is not a function
    at SIGNEXTEND (/home/circleci/project/ethreumjs-vm/dist/opFns.js:113:22)
    at runOp (/home/circleci/project/ethreumjs-vm/dist/runCode.js:199:27)
```

File: lib/opFns.js (hunk 1 above)
Comment (axic) [reply lib/opFns.js]: This part needs to be verified, not sure where this comment came from, but if state tests are passed with the current one that can only mean two things: a) it is correct or b) state tests are lacking.

**Comment by axic** [comment]
Still fails on
```
# file: randomStatetest470 test: randomStatetest470
not ok 1975 the state roots should match
  ---
    operator: equal
    expected: |-
      '803f04e36d462235dee12690809f33cb0020c810a6b421a46f97a89b5fade3b3'
    actual: |-
      '90d26a259711a6aa54fb777356285fc75e825c90f0fabf7ff3279c1bed2bb273'
    at: replenish (/home/circleci/project/ethereumjs-vm/node_modules/async/dist/async.js:1011:17)
  ...
```

**Comment by axic** [comment]
At this stage I'm not sure what fun the tests are playing with me. With various broken versions of this PR it is always the very same test case which fails. But broken versions are not detected..

**Comment by holgerd77** [comment]
@axic Can you update this?

**Comment by holgerd77** [comment]
@axic Can you do an update here?

**Comment by axic** [comment]
Down to again a single failure:
```
# file: randomStatetest303 test: randomStatetest303
not ok 1900 the state roots should match
  ---
    operator: equal
    expected: |-
      '341d35dcd68e14ed391c23300e89fb48d60096e473e4e8b84589df12381e8d0e'
    actual: |-
      '832f429e23eb9c488251f8506024e4cb172fc58c41ed620d1f1b572f5c3af88a'
    at: replenish (/home/circleci/project/ethereumjs-vm/node_modules/async/dist/async.js:1011:17)
  ...
```

**Comment by axic** [comment]
@holgerd77 @s1na woohoo this seems to be working now.

**Pull request merged**